pyvelox/complex.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "complex.h"
+#include "velox/vector/ComplexVector.h"
+
+#include <functional>
+
+namespace facebook::velox::py {
+
+using namespace velox;
+namespace py = pybind11;
+
+namespace {
+// Structure used to keep check on the number
+// of constituent elements. Attributes totalElements
+// and insertedElements keeps the length of the vector, and
+// the number of elements inserted during the operation
+// respectively.
+struct ElementCounter {
+  vector_size_t insertedElements =
+      0; // to track the elements already in the vector
+  vector_size_t totalElements = 0;
+  std::vector<ElementCounter> children;
+};
+} // namespace
+
+void checkOrAssignType(TypePtr& type, const std::function<TypePtr()>& func) {
+  if (type->kind() == TypeKind::UNKNOWN) {
+    type = func();
+  } else if (!(type->kindEquals(func()))) {
+    throw py::type_error(
+        "Cannot construct type tree, invalid variant for complex type");
+  }
+}
+
+template <TypeKind Kind>
+void setElementInFlatVector(
+    vector_size_t idx,
+    const variant& v,
+    VectorPtr& vector) {
+  using NativeType = typename TypeTraits<Kind>::NativeType;
+  auto asFlat = vector->asFlatVector<NativeType>();
+  asFlat->set(idx, NativeType{v.value<NativeType>()});
+}
+
+// Function responsible to construct the type for the passed variant
+// for a complex vector. It is supposed to run a recursive call with a
+// pre-instantiated TypePtr, the target variant and the counter. The
+// passed variant is checked for its data type, and for any complex
+// type involved, the function is called again. The counter here is used
+// to keep in track of the number of elements inserted and the number of
+// types of elements allowed if a complex vector is involved in the variant.
+void constructType(const variant& v, TypePtr& type, ElementCounter& counter) {
+  ++counter.totalElements;
+
+  if (v.isNull()) {
+    // since the variant is NULL, we can't infer the data type
+    // thus it maybe UNKNOWN or INVALID at this stage
+    // which implies further investigation is required
+    if (v.kind() != TypeKind::UNKNOWN && v.kind() != TypeKind::INVALID &&
```

Comment:
I think it should be an `and` condition, instead of an `or`. With `and`, it implies the situation of throwing an error when the type kind is none of the ones that are expected. If we instead make it an `or` condition, then say the kind is an unknown, that implies the first condition will be false, but the other two conditions (i.e. not being invalid and not being the type kind will be true) will be true, and thus the entire condition will be true and will throw the error, which should not be the case I guess?